### ISSUE TITLE
label_replace does not deduplicate series

### ISSUE THREAD

**Issue by jhesketh** [prometheus/prometheus#15114]
### What did you do?

```

load 20m
  series{label="a", idx="1"} 2 _
  series{label="a", idx="2"} _ 4

eval instant at 0 label_replace(series, "idx", "replaced", "idx", ".*")
  series{label="a", idx="replaced"} 2

eval instant at 20m label_replace(series, "idx", "replaced", "idx", ".*")
  series{label="a", idx="replaced"} 4

eval range from 0 to 20m step 20m label_replace(series, "idx", "replaced", "idx", ".*")
  series{label="a", idx="replaced"} 2 4

```

### What did you expect to see?

I expected the tests to pass.

### What did you see instead? Under which circumstances?

The range query fails with:
```error evaluating query "label_replace(series, \"idx\", \"replaced\", \"idx\", \".*\")" (line 11): vector cannot contain metrics with the same labelset```

**Comment by krajorama** [comment]
Hello from the bug scrub!

We think this is working as expected, label_replace isn't intended as a join operation.
Closing as such, but ping @juliusv if he has a different idea.

**Comment by juliusv** [comment]
Correct, `label_replace()` isn't intended to merge anything, it just plainly drops labels and the user is responsible for not producing any duplicate series. For merging, you have various aggregation operators: https://prometheus.io/docs/prometheus/latest/querying/operators/#aggregation-operators

**Comment by charleskorn** [comment]
I don't think this is desirable behaviour, and should be fixed: it means a query that works when run as separate instant queries at T=1 and T=2 fails when run as a range query from T=1 to T=2.

**Comment by juliusv** [comment]
@charleskorn Oh thanks so much for catching that, and apologies then - I had only read the message by @krajorama  fully and kinda only skimmed the initial report. That's definitely 100% a bug. Reopening.

I have tried to use `git bisect` to find the offending commit, but some intermediary commits didn't build so I couldn't test them (skipped them via `git bisect skip`). Here's what `git bisect` said in the end:

```
There are only 'skip'ped commits left to test.
The first bad commit could be any of:
babfcfdd910ef2d3e829ae05bac766c1c4e3952b
b3c15d2246e9652f08c2b6030b82f8aaa943849e
e7c77f7b4097eb4ddcde99aed2a9296870b79052
8fd96241ab04d829e198ad2f6f67e33f3bf5b564
9aa321d67236ab43c1c9ca69577fd361b87e2405
4a72607c4a59d51042a282d02b930e355b4ac033
ccf73266f03a7de3fbd7d150ee80ee9521ba17a1
0dbfd20b695a8dfa0ddc23e76fda69b6ffcbc4bb
11b27d5d229d7ee815bd7cfc77c6682b1f3eed0c
786e0e706cc90cb54f59f0c95704473a963f995f
We cannot bisect more!
```

These commits are all part of https://github.com/prometheus/prometheus/pull/13999 by @bboreham, and it seems like that PR aimed to only refactor the test framework itself. I wonder if the issue then is only in the refactored test framework, or really in the PromQL engine.

**Comment by juliusv** [comment]
Ah sorry, I think the git bisect only pointed at those commits because the whole directory with tests moved and with earlier versions my test command didn't even include the bug-producing expression test file anymore 🙈 So the error is likely somewhere else.

**Comment by juliusv** [comment]
Ok, I did a `git bisect` based on actual data and queries on a live Prometheus server and I tracked down this PR / commit:

https://github.com/prometheus/prometheus/pull/13536

@bboreham could you take a look?

**Comment by bboreham** [comment]
Yeah.  Do we really care?

**Comment by juliusv** [comment]
I think so, or are you suggesting to just leave it broken? I can see how someone might use an expression like the one in the test above to e.g. make multiple intermittent series look like the same one over time, or maybe drop some label for some other reason that is not required to differentiate series at one time step, but would be across multiple. There's maybe better ways to do that (aggregation), but at least it doesn't seem like an ultra arcane edge case that would never happen to anyone.

**Comment by beorn7** [comment]
Interesting discussion.

I was in the bug scrub where we concluded that the old behavior doesn't really make sense, but the discussion so far have convinced me that the old behavior actually makes more sense, but mostly from an academic perspective. I would still have my doubts how practically relevant this is, but as @charleskorn said, whenever a series of instant queries behaves different from the equivalent range query, we should pay attention if that's really necessary. (I was thinking the problem might be in rendering the API response payload, but that's not really a problem in this case.)

Maybe there is an easy fix to reinstate the old behavior without nixing the efficiency gains #13536…

**Comment by bboreham** [comment]
IMO it's only "broken" for an extremely specific situation, and aggregations are the way to make multiple series look like one.
It's been like this for a year and a half and no real users commented.

However, I realised that we can retain the performance benefits for most cases, either by having different behaviour only when the new label overwrites an old, or by doing the current thing first, checking if a clash arises, and then re-checking step by step if so.

**Comment by roidelapluie** [comment]
See https://github.com/prometheus/prometheus/pull/17703
